From e4aba0c3edb44700688b125e592f486808801d0e Mon Sep 17 00:00:00 2001
From: Motohiro Kosaki <mkosaki@redhat.com>
Date: Wed, 10 Sep 2014 16:16:02 -0400
Subject: [x86] kernel/setup: do reserve_crashkernel() after SRAT is parsed

Message-id: <1410365775-5132-8-git-send-email-mkosaki@redhat.com>
Patchwork-id: 93487
O-Subject: [RHEL7 PATCH 07/20] x86, acpi, crash, kdump: do reserve_crashkernel() after SRAT is parsed.
Bugzilla: 810042
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Larry Woodman <lwoodman@redhat.com>
RH-Acked-by: Rik van Riel <riel@redhat.com>

From: KOSAKI Motohiro <mkosaki@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=810042
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7914549
target RHEL version: 7.1
upstream status: merged
changes from upstream: none

commit fa591c4ae76ecbd4d26d7e8f65429d6d454554a6
Author: Tang Chen <tangchen@cn.fujitsu.com>
Date:   Tue Nov 12 15:08:07 2013 -0800

    x86, acpi, crash, kdump: do reserve_crashkernel() after SRAT is parsed.

    Memory reserved for crashkernel could be large.  So we should not allocate
    this memory bottom up from the end of kernel image.

    When SRAT is parsed, we will be able to know which memory is hotpluggable,
    and we can avoid allocating this memory for the kernel.  So reorder
    reserve_crashkernel() after SRAT is parsed.

    Signed-off-by: Tang Chen <tangchen@cn.fujitsu.com>
    Signed-off-by: Zhang Yanfei <zhangyanfei@cn.fujitsu.com>
    Acked-by: Tejun Heo <tj@kernel.org>
    Acked-by: Toshi Kani <toshi.kani@hp.com>
    Cc: Thomas Gleixner <tglx@linutronix.de>
    Cc: Ingo Molnar <mingo@elte.hu>
    Cc: "H. Peter Anvin" <hpa@zytor.com>
    Cc: Wanpeng Li <liwanp@linux.vnet.ibm.com>
    Cc: Thomas Renninger <trenn@suse.de>
    Cc: Yinghai Lu <yinghai@kernel.org>
    Cc: Jiang Liu <jiang.liu@huawei.com>
    Cc: Wen Congyang <wency@cn.fujitsu.com>
    Cc: Lai Jiangshan <laijs@cn.fujitsu.com>
    Cc: Yasuaki Ishimatsu <isimatu.yasuaki@jp.fujitsu.com>
    Cc: Taku Izumi <izumi.taku@jp.fujitsu.com>
    Cc: Mel Gorman <mgorman@suse.de>
    Cc: Michal Nazarewicz <mina86@mina86.com>
    Cc: Minchan Kim <minchan@kernel.org>
    Cc: Rik van Riel <riel@redhat.com>
    Cc: Johannes Weiner <hannes@cmpxchg.org>
    Cc: Kamezawa Hiroyuki <kamezawa.hiroyu@jp.fujitsu.com>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Signed-off-by: KOSAKI Motohiro <mkosaki@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kernel/setup.c b/arch/x86/kernel/setup.c
index f2a890e..8af4b74 100644
--- a/arch/x86/kernel/setup.c
+++ b/arch/x86/kernel/setup.c
@@ -1193,8 +1193,6 @@ void __init setup_arch(char **cmdline_p)
  acpi_initrd_override((void *)initrd_start, initrd_end - initrd_start);
 #endif
 
- reserve_crashkernel();
-
  vsmp_init();
 
  io_delay_init();
@@ -1215,6 +1213,13 @@ void __init setup_arch(char **cmdline_p)
  early_acpi_boot_init();
 
  initmem_init();
+
+ /*
+  * Reserve memory for crash kernel after SRAT is parsed so that it
+  * won't consume hotpluggable memory.
+  */
+ reserve_crashkernel();
+
  memblock_find_dma_reserve();
 
 #ifdef CONFIG_KVM_GUEST
-- 
1.7.1